### DISCUSSION TITLE
Games using Unity plugin Rewired do not get blocked by HidHide

### DISCUSSION THREAD

**Discussion by Yohoki** [nefarius/HidHide#38] [Q&A] [answered]
Update: Turns out, none of these games were leaking inputs. I just didn't know you needed to block your controller twice if you use it via USB cable AND BT.

**Comment by nefarius** [comment]
Kindly don't open a bug report with literally no useful information attached, I took liberty to move it to a Discussion, thank you.

**Reply by Yohoki** [reply]
Sorry about that. What kind of info can I provide to help? :)

**Reply by nefarius** [reply]
A very important detail for example is the kind of input API those games use. DirectInput and HID-API for example are supposed to be blocked perfectly, once XInput enters the scene it gets complicated. We currently have reports that that is broken in the UI client because it does not tell the driver the right information to block the XInput device as well. So this may be the case you're encountering since most games rely on XI nowadays anyway.

**Reply by Yohoki** [reply]
Ah. I see. That's only based on which controller is being used, right? It didn't occur to me that the controller used might be misunderstood. I'm using a DS4, so it should be DirectInput that the game is picked up. If that's what you're meaning.

If you're meaning I need to figure out which API is programmed into the game, I would need to know how to do that. I wouldn't even know where to start to figure that out.

**Reply by nefarius** [reply]
Yes, it heavily depends on the controller and indeed isn't an easy task because unfortunately it isn't common for many games or engines to publicly document what input APIs they support and there are unfortunately a few that arose over time. The DS4 should be an easy example, it is accessible via DirectInput, Raw Input and HID API. Welcome to the messy world of gaming input peripherals!

**Reply by Yohoki** [reply]
So then for these two games, assuming DInput and HID-API are actually 100% blocked, that would make it have to be Raw Input that is being used. But then again I also know that I'm in the wonderfully messy world of coding and probably shouldn't assume that either of those are 100% blocked! XD

But other than the guess work, there's really no way of knowing which API is being used without inspecting the code or getting a dev to say.... 

OH! I just found a site claiming Two Point Hospital was made with Rewired (A controller mapping helper? I think?). The info for this program says "On Windows the controller must be compatible with Raw Input, Direct Input, or XInput"... So, since it's a DS4 controller, it has to be either Raw or DInput that is to blame. :D I did a thing! XD

**Reply by nefarius** [reply]
It *is* a mess and difficult task. I myself am currently "investigating" the game [Blasphemous](https://store.steampowered.com/app/774361/Blasphemous/) and try my new emulation approach on it; this game (or its engine rather) for example uses DirectInput (or Raw Input, HID-API, doesn't really matter) **and** XInput at the same time, which causes weirdness when the XInput controller also is a HID-compliant device picked up through DirectInput. It causes issues like doubled inputs or rapidly switching between keyboard and controller layouts. Fun!

**Reply by Yohoki** [reply]
Using D4W to emulate m+kb controls on a ds4, I can say I'm very familiar with the keyboard/controller layouts flashing! And seeing the two controllers pop up in games. XD

Thankfully, I haven't seen that near as much as I used to now that HidHide is a thing. It's fixed so many issues, so thanks for the hard work you've done so far!

I also got a chance to look into the Two Point Hospital game a bit more... I tried opening the DLL files for the game in Ghidra, and I'm noticing a lot of code calling things like AxisRaw and ButtonRaw. I thought that maybe the "AxisRaw" might be referring to data before applying deadzone values and things... but I also see a "Rewired.Data.Mapping.HardwareJoystickMap.Platform_RawOrDirectInput.Axis_Base" . So, it doesn't help narrow down between DInput and RawInput.... but it at least confirms that it's one of these two that are being leaked.

**Reply by Yohoki** [reply]
I got a chance to take a look a bit deeper into this. It seems I actually have a lot of games with Rewired in my library.... So, I haven't had a chance to test all of them yet, but I've added them to the list anyway for now. I did check several out, though.

I also noticed that there are several Rewired DLLs. So, I need to check if having any of them circumvents HidHide or just a few of them. So far I've found these DLLs in my library.
- Rewired_Core.dll (Confirmed - 'A Blind Legend' uses only this dll and bypasses HidHide.)
- Rewired_Windows_Lib.dll / Rewired_Windows.dll (Unsure - Most Games use one of these along with the Core DLL.)
- Rewired.dll (Unsure - I'm only seing it used alongside Core DLL)
- Rewired_DirectInput.dll (Unsure - Raw Input still used if the Core DLL is present, but I have no games that only use this DLL by itself)
- TPRuntime_RewiredInternal.dll / TPRuntime_RewiredIntegration.dll (Unsure - 'Empire of Sin' only?)

**Reply by Yohoki** [reply]
Also got a chance to look at Breakwaters again and it doesn't seem to be using Rewired. I did a good search through the folders and can't seem to find any files that look like they'd tell me what was being used either... But, I was able to open unity explorer up inside the game and took some looking around. The best thing I can find is a bunch of mentions to Unity's InputSystem, which is... well... it's input system? XD

The input system seems to sopport, like, everything. But more importantly, I see a few mentions of "Low Level" and "Low Level API" on the documentation... I'm assuming Low Level means Raw?
![image](https://user-images.githubusercontent.com/44595371/126857184-20e15d33-74a3-485d-9792-c6609252df4a.png)

So, if that's the case, this game is also using Raw Input API. I've solved the case! Yay!

But I still don't know how to fix any of that. lol But, at least now we know why these games aren't behaving right with HidHide: Raw Input APIs.

**Comment by Yohoki** [comment, marked as answer]
Son of a gun....

It turns out, when you get one of those fancy new fangled bluetooth dongles.... you also need to block your controller on BT as well as USB connection.... (｡･_･｡)

**Reply by nefarius** [reply]
That... Was always the case. Every protocol uses its own device object that is unique and independent from the USB device object.

**Reply by Yohoki** [reply]
(｡･_･｡)
Yep. It sure was, wasn't it.
